### PR TITLE
feat(runtime): abstract Claude Code behind pluggable agent runtime

### PR DESCRIPTION
**Pull request by ifireball**
## Summary

- Add `internal/runtime` with a `Runtime` interface and `ClaudeRuntime` as the default implementation.
- Runner uses `BootstrapInput` and optional `ClaudeHooksBootstrap` (via harness adapter) instead of calling Claude Code directly from `cli/run`.
- Narrow `security.ClaudeSandboxHooks` for hook generation; host-side scans stay in the runner.

## Test plan

- [ ] `go test ./internal/runtime/... ./internal/security/...`
- [ ] `go test ./internal/cli/ -run 'TestNewHarnessBootstrap|TestBuildRunCommand|TestBuildScanContext'`
- [ ] Smoke `fullsend run` against a harness agent in CI or locally

Made with [Cursor](https://cursor.com)

### CODE REVIEW

File: internal/runtime/claude.go [hunk 1]
```diff
@@ -0,0 +1,399 @@
...
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
```
Comment (waynesun09) [inline internal/runtime/claude.go:11]: **HIGH — `runtime` package imports `harness`, creating inverted dependency**

The `runtime` package imports `internal/harness` solely for `harness.BoolDefault()` (one call site in `installClaudeHooks`). This creates a `runtime → harness` dependency that inverts the intended abstraction layers — `runtime` is meant to be a lower-level abstraction that `cli` delegates to, so having it reach back into `harness` (the configuration/definition layer) is a layering violation.

The `security` package already has its own unexported `boolDefault` copy in `hooks.go:195`.

**Suggestion:** Either inline the one-liner (`if b == nil { return def }; return *b`), use it from the `security` package, or extract to a shared `internal/util` package. The single call site does not justify the package-level coupling.

File: internal/runtime/claude_test.go [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+		"claude-plugins-official", "1.0.0", "/tmp/claude-config",
+	)
+	require.NoError(t, err)
+	require.Len(t, entries, 4)
+
+	var settings map[string]any
+	require.NoError(t, json.Unmarshal(entries[3].data, &settings))
+	enabled := settings["enabledPlugins"].(map[string]any)
+	assert.Len(t, enabled, 0)
+}
```
Comment (waynesun09) [inline internal/runtime/claude_test.go:285]: **HIGH — 8 test cases dropped without replacement, including security-relevant quote escaping tests**

19 test functions were deleted from `run_test.go` but only 11 were added here. Missing tests:

**Security-relevant (must port):**
- `PluginDirEscapesQuotes` — verifies single-quote escaping in plugin paths prevents shell injection
- `DebugEscapesQuotes` — verifies quote escaping in debug filter flag
- `NoDoubleSpaces` — 6-case table test preventing command formatting regressions

**Edge case coverage (should port):**
- `MultiplePluginDirs`, `NoPlugins`, `DebugDisabled`
- `MultiplePlugins`, `NoLspJSON`, `InvalidLspJSON`, `EmptyLspJSON`, `ConfigStructure`

The function logic is identical — these tests only need the `testRunCommand` helper wrapper to port.

File: internal/runtime/bootstrap.go [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+	AgentPath() string
+	SkillDirs() []string
+	PluginDirs() []string
+}
+
+// ClaudeHooksBootstrap is an optional extension for Claude Code sandbox tool hooks.
+// ClaudeRuntime.Bootstrap type-asserts for this; other runtimes ignore it.
+type ClaudeHooksBootstrap interface {
+	ClaudeSandboxHooks() security.ClaudeSandboxHooks
+}
```
Comment (waynesun09) [inline internal/runtime/bootstrap.go]: **HIGH — `ClaudeHooksBootstrap` in shared `runtime` package couples the portable interface layer to Claude**

`bootstrap.go` imports `internal/security` for the `ClaudeHooksBootstrap` interface. This puts a Claude-specific extension type with a `security` dependency in the shared runtime interface file. As more runtimes are added, this file would accumulate runtime-specific extension interfaces and their imports, defeating the purpose of the abstraction.

The `BootstrapInput` interface above is clean and portable — this extension breaks that pattern.

**Suggestion:** Move `ClaudeHooksBootstrap` to a separate file (e.g., `claude_bootstrap.go`) within the runtime package, or into the `security` package itself. The `ClaudeRuntime.Bootstrap` method can type-assert against an interface defined closer to the Claude implementation rather than in the shared interface file.

File: internal/cli/bootstrap_scan.go
Comment (waynesun09) [inline internal/cli/bootstrap_scan.go:1]: **MEDIUM — No tests for 99 lines of security-critical scanning code**

`scanRuntimeContent`, `scanAgentFile`, `scanSkillDir`, and `scanPluginDir` are newly extracted and easily testable, but no `bootstrap_scan_test.go` exists. These functions gate what enters the sandbox — a bug here could silently skip blocking malicious content.

The PR's test plan lists `TestBuildScanContext` but that tests the in-sandbox scan, not this host-side scan.

**Suggestion:** Add `bootstrap_scan_test.go` covering:
1. Agent file with critical findings + `failClosed=true` → returns error
2. Agent file with critical findings + `failClosed=false` → logs warning, returns nil
3. Skill dir with no SKILL.md + `failClosed=true` → warns but succeeds
4. Plugin dir with critical findings in `plugin.json` → blocks in fail-closed mode

These are pure filesystem tests that work with `t.TempDir()` — no sandbox required.

File: internal/cli/run.go [hunk 4]
```diff
@@ -337,9 +336,25 @@ func runAgent(agentName, fullsendDir, outputBase, targetRepo, fullsendBinary str
...
 	// 7. Bootstrap sandbox.
+	rt := agentruntime.Default()
 	bootstrapStart := time.Now()
 	printer.StepStart("Bootstrapping sandbox")
-	if err := bootstrapSandbox(sandboxName, repoDir, fullsendBinary, h); err != nil {
+	boot := newHarnessBootstrap(h, sandboxName)
+	if h.SecurityEnabled() {
+		if err := scanRuntimeContent(boot, h.FailModeClosed()); err != nil {
+			printer.StepFail("Failed to bootstrap sandbox")
+			return err
```
Comment (waynesun09) [inline internal/cli/run.go:398]: **MEDIUM — TOCTOU: scan and upload are no longer atomic per file**

The old code interleaved scanning and uploading per file (scan agent → upload agent → scan skill → upload skill). The new code separates these into phases: `scanRuntimeContent` scans all files first (here), then `rt.Bootstrap` uploads them all later (~13 lines below). Between the scan and the upload, the host filesystem files could theoretically be modified.

Practical risk is low (requires concurrent host write access during pipeline execution), but this is a defense-in-depth regression from the previous design where scan-then-upload was immediate per file.

**Suggestion:** Document this as an accepted risk (host filesystem is controlled by the runner, not the agent), or read file content during scan and pass in-memory bytes to the upload step so the same bytes are scanned and uploaded.

**Note:** The new ordering (all warnings before any upload) is arguably better for UX — you get the full scan picture before any sandbox modification.

File: internal/runtime/runtime.go [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+	ConfigDir() string
+	EnvExports() []string
+	Bootstrap(input BootstrapInput) error
+	Run(params RunParams, printer *ui.Printer, start time.Time, metrics *RunMetrics) (exitCode int, err error)
+	ClearIterationArtifacts(sandboxName string) error
+	ExtractTranscripts(sandboxName, agentLabel, outputDir string) error
+	ExtractDebugLog(sandboxName, localPath, debug string) error
+	ParseTranscriptErrors(transcriptDir string) []TranscriptError
+	EmitTranscriptErrors(w io.Writer, summaries []TranscriptError)
+}
```
Comment (waynesun09) [inline internal/runtime/runtime.go:43]: **MEDIUM — `Runtime` interface is too wide (10 methods) for MVP with single implementation**

Several methods are not "runtime" responsibilities:
- `ParseTranscriptErrors` / `EmitTranscriptErrors` — pure functions that read local files and format GHA annotations. No sandbox interaction. Another runtime would implement identical `EmitTranscriptErrors` (same GHA format) but different `ParseTranscriptErrors` (different transcript format).
- `ExtractDebugLog` — Claude-specific (`claude-debug.log` path, `--debug` CLI flag). Other runtimes may not have debug logs at all.
- `ClearIterationArtifacts` — hardcodes `output/*` and `*.jsonl` glob patterns.

A second runtime would need to implement all 10 methods even if most are no-ops.

Also: `SandboxName` is passed through 3 different mechanisms — `BootstrapInput.SandboxName()`, `RunParams.SandboxName`, and direct `string` parameter. Inconsistent for implementors.

**Suggestion:** Split into core `Runtime` (Name, ConfigDir, EnvExports, Bootstrap, Run, ClearIterationArtifacts) and optional `TranscriptExtractor` interface. Or make `ParseTranscriptErrors`/`EmitTranscriptErrors` standalone functions since they have no sandbox interaction. Acceptable for MVP, but add a TODO noting the interface needs refinement before a second runtime.

File: internal/security/claude_hooks.go [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+
+import (
+	"github.com/fullsend-ai/fullsend/internal/harness"
+)
+
+// ClaudeSandboxHooks configures Claude Code PreToolUse/PostToolUse security hooks.
+// A nil SandboxHooks field uses the same defaults as an unset harness security block.
+type ClaudeSandboxHooks struct {
+	SandboxHooks *harness.SandboxHooks
+}
```
Comment (waynesun09) [inline internal/security/claude_hooks.go:11]: **MEDIUM — `ClaudeSandboxHooks` exposes `harness.SandboxHooks` as a public mutable field**

Two issues:

1. **Leaks harness internals:** The exported `SandboxHooks *harness.SandboxHooks` field lets any code with a `ClaudeSandboxHooks` value access and depend on the full harness struct, including fields like `Tirith`, `SSRFPreTool`, etc. The `runtime` package accesses it directly (`hooks.SandboxHooks.Tirith`), coupling runtime to harness internals through the abstraction.

2. **Inconsistent access pattern:** Some callers use the exported field directly (e.g., `installClaudeHooks` in `claude.go`: `sh := hooks.SandboxHooks`), while the hook enable/disable functions in `hooks.go` use the unexported `sandboxHooks()` method. Both do the same thing — `return c.SandboxHooks`.

**Suggestion:** Make the field unexported and use the `sandboxHooks()` accessor consistently. Or extract the needed config values into explicit fields (`TirithFailOn string`, `TirithRequired bool`) so the runtime doesn't need to know about harness types at all.

File: internal/sandbox/sandbox.go [hunk 7]
```diff
@@ -536,6 +536,8 @@ func collectPodmanLogs(sandboxName string) string {
 
 // ExtractTranscripts copies Claude transcript files (.jsonl) from the sandbox
 // to a local output directory.
+//
+// Deprecated: use runtime.ClaudeRuntime.ExtractTranscripts via runtime.Default().
```
Comment (waynesun09) [inline internal/sandbox/sandbox.go]: **MEDIUM — Duplicate `ExtractTranscripts` — deprecated copy left with no remaining callers**

`ClaudeRuntime.ExtractTranscripts` in `runtime/claude.go` is a near-exact copy of this function. The only code caller was in `run.go`, which this PR switches to `rt.ExtractTranscripts`. Having two copies of the same traversal-safety logic (using `os.OpenRoot`) means any future security fix must be applied in two places.

**Suggestion:** Delete this function entirely since no code callers remain after this PR. Git history preserves it if needed. Alternatively, have the runtime version delegate to this function to avoid duplication, but deletion is cleaner.

**Review by waynesun09 — COMMENTED** [review]
## Review Squad Report — 7 agents, 14 findings after dedup, 4 false positives removed

**Agents:** 2x claude-coder, 2x claude-researcher, 2x gemini-code-review, 1x cursor-code-review

Clean mechanical refactor that successfully abstracts Claude Code behind a `Runtime` interface. The architecture and code movement are sound — no behavioral bugs detected.

### Summary of findings (HIGH: 3, MEDIUM: 6, LOW: 4, INFO: 1)

**High:**
1. `runtime → harness` dependency inversion via `BoolDefault` (5-agent consensus)
2. 8 test cases dropped without replacement, including shell-injection-preventing quote escaping tests (5-agent consensus)
3. `ClaudeHooksBootstrap` in shared interface file couples portable layer to Claude (2 agents)

**Medium:**
4. No tests for 99 lines of security-critical `bootstrap_scan.go` (5-agent consensus)
5. TOCTOU regression — scan and upload no longer atomic per file (3 agents)
6. `Runtime` interface too wide at 10 methods + inconsistent `SandboxName` passing (5-agent consensus)
7. `ClaudeSandboxHooks` exports mutable `harness.SandboxHooks` pointer (5-agent consensus)
8. Duplicate `ExtractTranscripts` — deprecated copy with no remaining callers (4 agents)

See inline comments for details and suggestions.

File: internal/runtime/runtime.go [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+
+// Default returns the configured agent runtime (Claude Code today).
+func Default() Runtime {
+	return ClaudeRuntime{}
+}
```
Comment (ralphbean) [inline internal/runtime/runtime.go:55]: **[important]** With this PR we're making the decision that runtimes are pluggable — which is great, it lines up with #1260, #579, #70. But I think we should be explicit about the security surface of each runtime before people start choosing between them.

The `Runtime` interface has no contract around security hooks. `ClaudeHooksBootstrap` is an optional type assertion in `ClaudeRuntime.Bootstrap()`, so a future runtime can satisfy the interface without installing any security protections at all — tirith, SSRF detection, canary tokens, secret redaction, unicode normalization, all silently absent.

I think we need a `docs/runtimes.md` (or similar) that documents which runtimes provide which security features, what scanners they depend on from the sandbox image, and any other trade-offs. Even with Claude as the only runtime today, writing that table sets the template so the next runtime addition forces the author to fill in the gaps. Something like:

| Feature | Claude Code | (future) |
|---------|------------|----------|
| tirith (Bash scanning) | ✓ | ? |
| SSRF pre-tool | ✓ | ? |
| Canary token detection | ✓ | ? |
| Secret redaction | ✓ | ? |
| Unicode normalization | ✓ | ? |
| Tool allowlist | opt-in | ? |
| Prompt injection (DeBERTa) | ✓ (host-side) | ? |

Does that seem reasonable?

File: internal/runtime/claude_test.go (hunk 2 above)
Comment (ralphbean) [inline internal/runtime/claude_test.go:285]: **[important]** A few tests from `run_test.go` didn't make the trip over. The ones I'd most want to see ported are the shell-escaping edge cases — `PluginDirEscapesQuotes` and `DebugEscapesQuotes` test defense-in-depth against command injection, and the malformed-input tests (`InvalidLspJSON`, `EmptyLspJSON`, `NoLspJSON`) cover graceful handling of broken plugin configs. The `NoDoubleSpaces` table test (6 subtests) also caught a real class of command-construction bug.

Full list of what I think is missing:
- `PluginDirEscapesQuotes`, `DebugEscapesQuotes` (shell injection defense)
- `InvalidLspJSON`, `EmptyLspJSON`, `NoLspJSON` (malformed input resilience)
- `NoDoubleSpaces` (6 subtests, command construction correctness)
- `MultiplePluginDirs`, `DebugDisabled`, `MultiplePlugins`, `ConfigStructure`

The security-relevant ones are the priority — would you be up for porting at least those?

File: internal/runtime/runtime.go [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+
+import (
+	"io"
+	"sync/atomic"
+	"time"
+
+	"github.com/fullsend-ai/fullsend/internal/ui"
+)
+
+const claudeDebugLog = "claude-debug.log"
```
Comment (ralphbean) [inline internal/runtime/runtime.go]: **[moderate]** `claudeDebugLog` is only used by `ClaudeRuntime.ExtractDebugLog()` and `buildRunCommand()` in `claude.go`. Could we move it there? A different runtime would have its own debug artifact name.

**Review by ralphbean — CHANGES_REQUESTED** [review]
I think this needs a few changes before we can merge. See inline comments.

File: internal/runtime/claude.go (hunk 1 above)
Comment (ifireball) [reply internal/runtime/claude.go:11]: Fixed in e14d231c: removed the `runtime`→`harness` import. Tirith env wiring now uses `security.ClaudeSandboxHooks.TirithFailOn()` / `TirithRequired()` so `runtime` does not touch harness types.

File: internal/runtime/claude_test.go (hunk 2 above)
Comment (ifireball) [reply internal/runtime/claude_test.go:285]: Fixed in e14d231c: ported the missing command/plugin tests to `internal/runtime/claude_test.go` (quote escaping, NoDoubleSpaces table, plugin LSP edge cases, etc.).

File: internal/runtime/bootstrap.go (hunk 3 above)
Comment (ifireball) [reply internal/runtime/bootstrap.go]: Fixed in e14d231c: moved `ClaudeHooksBootstrap` to `internal/runtime/claude_bootstrap.go`; `bootstrap.go` stays portable (`BootstrapInput` only).

File: internal/cli/bootstrap_scan.go
Comment (ifireball) [reply internal/cli/bootstrap_scan.go:1]: Fixed in e14d231c: added `internal/cli/bootstrap_scan_test.go` with the four cases you listed (fail-closed/open, missing SKILL.md, plugin critical).

File: internal/cli/run.go (hunk 4 above)
Comment (ifireball) [reply internal/cli/run.go:398]: Addressed: comment in `run.go` documents accepted risk (runner-owned host FS). Follow-up tracked in fullsend-ai/fullsend#1801 for checksum/in-memory hardening.

File: internal/runtime/runtime.go (hunk 5 above)
Comment (ifireball) [reply internal/runtime/runtime.go:43]: Fixed in e14d231c: split `TranscriptHandler` from core `Runtime`; `Default()` returns a `Backend` bundling both. `run.go` uses `tx` for transcript/debug ops.

File: internal/security/claude_hooks.go (hunk 6 above)
Comment (ifireball) [reply internal/security/claude_hooks.go:11]: Fixed in e14d231c: `ClaudeSandboxHooks` field is unexported; hook enablement uses `sandboxHooks()` internally; runtime uses `TirithFailOn`/`TirithRequired` only.

File: internal/sandbox/sandbox.go (hunk 7 above)
Comment (ifireball) [reply internal/sandbox/sandbox.go]: Fixed in e14d231c: deleted deprecated `sandbox.ExtractTranscripts`; sole implementation is `ClaudeRuntime.ExtractTranscripts`.

File: internal/runtime/runtime.go (hunk 8 above)
Comment (ifireball) [reply internal/runtime/runtime.go:55]: Done in e14d231c: added `docs/runtimes.md` with the security feature matrix and interface contracts; linked from `docs/architecture.md`.

File: internal/runtime/claude_test.go (hunk 2 above)
Comment (ifireball) [reply internal/runtime/claude_test.go:285]: Done — same port as the waynesun09 thread: shell-escaping and malformed plugin tests are in `claude_test.go` (e14d231c).

File: internal/runtime/runtime.go (hunk 9 above)
Comment (ifireball) [reply internal/runtime/runtime.go]: Fixed in e14d231c: `claudeDebugLog` constant lives in `claude.go` only (removed from `runtime.go`).

File: internal/cli/bootstrap_scan.go [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+	return nil
+}
+
+func scanAgentFile(pipeline *security.Pipeline, agentPath string, failClosed bool) error {
+	content, err := os.ReadFile(agentPath)
+	if err != nil {
+		if failClosed {
+			return fmt.Errorf("cannot scan agent definition %q: %w", agentPath, err)
+		}
+		fmt.Fprintf(os.Stderr, "WARNING: could not read agent definition %q for scan: %v\n", agentPath, err)
```
[the remark this answers is not in the document]
Comment (ifireball) [reply internal/cli/bootstrap_scan.go:52]: Fixed in ae799210: `scanRuntimeContent` returns a clear error when `AgentPath()` is empty (test: `TestScanRuntimeContent_EmptyAgentPath`).

File: internal/runtime/claude.go [hunk 11]
```diff
@@ -0,0 +1,407 @@
...
+}
+
+func buildRunCommand(params RunParams) string {
+	envFile := sandbox.SandboxWorkspace + "/.env"
+	safe := strings.ReplaceAll(params.AgentBaseName, "'", "'\\''")
+
+	parts := []string{
+		fmt.Sprintf("cd %s && . %s && claude", params.RepoDir, envFile),
+		"--print",
+		"--verbose",
```
[the remark this answers is not in the document]
Comment (ifireball) [reply internal/runtime/claude.go:190]: Fixed in ae799210: comment above `installClaudeHooks` documents the two intentional settings.json paths (plugin marketplace under `CLAUDE_CONFIG_DIR` vs security hooks under workspace `.claude/`).

File: internal/cli/run.go [hunk 12]
```diff
@@ -385,9 +384,27 @@ func runAgent(ctx context.Context, agentName, fullsendDir, outputBase, targetRep
...
 	bootstrapStart := time.Now()
 	printer.StepStart("Bootstrapping sandbox")
-	if err := bootstrapSandbox(sandboxName, repoDir, fullsendBinary, h); err != nil {
+	boot := newHarnessBootstrap(h, sandboxName)
+	if h.SecurityEnabled() {
+		// Scan all runtime content before upload so warnings surface together.
+		// Host files could change between scan and upload; the runner owns the host FS here.
+		if err := scanRuntimeContent(boot, h.FailModeClosed()); err != nil {
+			printer.StepFail("Failed to bootstrap sandbox")
+			return err
```
[the remark this answers is not in the document]
Comment (ifireball) [reply internal/cli/run.go:398]: Acknowledged — same as the TOCTOU thread: accepted risk documented in `run.go`, follow-up fullsend-ai/fullsend#1801.

File: internal/runtime/claude.go [hunk 13]
```diff
@@ -0,0 +1,404 @@
...
+	configDir := r.ConfigDir()
+
+	mkdirCmd := fmt.Sprintf("mkdir -p %s/agents %s/skills %s/hooks %s/plugins",
+		configDir, configDir, configDir, configDir)
+	if _, _, _, err := sandbox.Exec(sandboxName, mkdirCmd, 10*time.Second); err != nil {
+		return fmt.Errorf("creating runtime config dirs: %w", err)
+	}
+
+	if err := sandbox.Upload(sandboxName, input.AgentPath(),
+		fmt.Sprintf("%s/agents/", configDir)); err != nil {
```
Comment (maruiz93) [inline internal/runtime/claude.go:48]: **[moderate]** `Bootstrap()` passes `input.AgentPath()` directly to `sandbox.Upload` with no empty-string guard. When `SecurityEnabled()` is false, `scanRuntimeContent` is skipped — and that's the only place that validates `agentPath != ""`. An empty path would produce a confusing sandbox upload error rather than a clear validation message.

Practical risk is low (harness validation prevents empty agent paths upstream), but the runtime should fail clearly if called with bad input:
```go
if input.AgentPath() == "" {
    return fmt.Errorf("agent path is required")
}
```

File: internal/runtime/sanitize.go [hunk 14]
```diff
@@ -0,0 +1,29 @@
...
+	"regexp"
+	"strings"
+)
+
+// ansiEscRe matches ANSI CSI sequences, OSC sequences, and charset designators.
+var ansiEscRe = regexp.MustCompile(`\x1b\[[0-9;]*[a-zA-Z]|\x1b\][^\x07]*\x07|\x1b[()][A-Z0-9]`)
+
+// SanitizeOutput strips ANSI escape sequences, control characters, and GHA
+// workflow command markers from untrusted sandbox output.
+func SanitizeOutput(s string) string {
```
Comment (maruiz93) [inline internal/runtime/sanitize.go]: **[moderate]** `SanitizeOutput` is exported but all callers are within the `runtime` package (`claude_progress.go`, `claude_transcript.go`). Exporting a security-relevant function unnecessarily widens the API surface. Consider reverting to lowercase `sanitizeOutput`.

File: docs/runtimes.md [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+
+Harness `security.fail_mode` controls whether critical findings **block** the run (`closed`, default) or **warn** and continue (`open`). This applies to host scans, sandbox `scan context`, and host-side runtime content scan alike.
+
+### Runtime interface contract
+
+| Interface | Responsibility |
+|-----------|----------------|
+| `runtime.Runtime` | Name, config dir, env exports, bootstrap, run loop, per-iteration artifact cleanup |
+| `runtime.BootstrapInput` | Portable paths for agent/skills/plugins to upload |
+| `runtime.ClaudeHooksBootstrap` | Optional — Claude-only sandbox security hooks |
```
Comment (rh-hemartin) [inline docs/runtimes.md:34]: Shouldn't this be called `Hooks` in case other runtimes support some kind of hooks?

File: internal/cli/run.go [hunk 16]
```diff
@@ -723,13 +747,9 @@ func runAgent(ctx context.Context, agentName, fullsendDir, outputBase, targetRep
...
 
-func bootstrapSandbox(sandboxName, repoDir, fullsendBinary string, h *harness.Harness) error {
-	// Create workspace structure and Claude config dir for transcripts.
-	// Agent and skill definitions go in CLAUDE_CONFIG_DIR so `claude --agent`
-	// finds them regardless of the repo's own .claude/ directory. When
-	// CLAUDE_CONFIG_DIR is set, Claude uses it instead of ~/.claude/.
-	mkdirCmd := fmt.Sprintf("mkdir -p %s/agents %s/skills %s/hooks %s/plugins %s/bin %s/.env.d %s/.security %s %s/.claude/hooks",
-		sandbox.SandboxClaudeConfig, sandbox.SandboxClaudeConfig, sandbox.SandboxClaudeConfig, sandbox.SandboxClaudeConfig, sandbox.SandboxWorkspace, sandbox.SandboxWorkspace, sandbox.SandboxWorkspace, sandbox.SandboxClaudeConfig, sandbox.SandboxWorkspace)
+func bootstrapCommon(sandboxName, repoDir, fullsendBinary string, h *harness.Harness) error {
+	mkdirCmd := fmt.Sprintf("mkdir -p %s/bin %s/.env.d %s/.security %s/.claude/hooks",
```
Comment (rh-hemartin) [inline internal/cli/run.go]: Shouldn't we have a generic hooks folder and then link it somehow at Claude config level?

File: internal/runtime/claude.go [hunk 17]
```diff
@@ -0,0 +1,404 @@
...
+	if waitErr != nil && execCmd.ProcessState == nil {
+		return exitCode, fmt.Errorf("openshell exec failed: %w", waitErr)
+	}
+
+	return exitCode, nil
+}
+
+func (ClaudeRuntime) ClearIterationArtifacts(sandboxName string) error {
+	clearCmd := fmt.Sprintf("rm -rf %s/output/* %s/*.jsonl",
+		sandbox.SandboxWorkspace, sandbox.SandboxClaudeConfig)
```
Comment (rh-hemartin) [inline internal/runtime/claude.go]: I think this should use `r.ConfigDir()` and probably we need `r.WorkspaceDir()` as well.

File: internal/runtime/claude.go [hunk 18]
```diff
@@ -0,0 +1,404 @@
...
+		return fmt.Errorf("creating output dir: %w", err)
+	}
+
+	root, err := os.OpenRoot(outputDir)
+	if err != nil {
+		return fmt.Errorf("opening output root: %w", err)
+	}
+	defer root.Close()
+
+	configDir := sandbox.SandboxClaudeConfig
```
Comment (rh-hemartin) [inline internal/runtime/claude.go]: I think we want `r.ConfigDir()` here

File: internal/runtime/claude.go [hunk 19]
```diff
@@ -0,0 +1,404 @@
...
+		}
+	}
+
+	hooksInput, ok := input.(ClaudeHooksBootstrap)
+	if !ok {
+		return nil
+	}
+	return installClaudeHooks(sandboxName, hooksInput.ClaudeSandboxHooks())
+}
+
```
[the remark this answers is not in the document]
Comment (ifireball) [reply internal/runtime/claude.go:81]: Fixed in 8bf10e4c: `Bootstrap()` now requires a non-empty agent path and skips empty skill/plugin entries, matching `scanRuntimeContent`.

File: internal/cli/run.go [hunk 20]
```diff
@@ -385,9 +384,29 @@ func runAgent(ctx context.Context, agentName, fullsendDir, outputBase, targetRep
...
+	rt := backend.Runtime
+	tx := backend.Transcripts
 	bootstrapStart := time.Now()
 	printer.StepStart("Bootstrapping sandbox")
-	if err := bootstrapSandbox(sandboxName, repoDir, fullsendBinary, h); err != nil {
+	boot := newHarnessBootstrap(h, sandboxName)
+	if h.SecurityEnabled() {
+		// Scan all runtime content before upload so warnings surface together.
+		// Host files could change between scan and upload; the runner owns the host FS here.
+		if err := scanRuntimeContent(boot, h.FailModeClosed()); err != nil {
```
[the remark this answers is not in the document]
Comment (ifireball) [reply internal/cli/run.go:396]: Acknowledged — accepted risk with comment in `run.go`; tracked in fullsend-ai/fullsend#1801.

File: internal/runtime/claude.go (hunk 13 above)
Comment (ifireball) [reply internal/runtime/claude.go:48]: Fixed in 8bf10e4c: `Bootstrap()` returns `agent path is required` before any sandbox upload (see `TestBootstrap_EmptyAgentPath`).

File: internal/runtime/sanitize.go (hunk 14 above)
Comment (ifireball) [reply internal/runtime/sanitize.go]: Fixed in 8bf10e4c: reverted to unexported `sanitizeOutput` (package-local only).

File: docs/runtimes.md (hunk 15 above)
Comment (ifireball) [reply docs/runtimes.md:34]: Kept `ClaudeHooksBootstrap` intentionally: it is an optional type-assert extension for Claude security hooks, not a generic hook contract. A future runtime would add its own `*FooHooksBootstrap` interface rather than a shared `Hooks` name that implies parity across backends. Happy to add a naming note in `docs/runtimes.md` if you prefer.

File: internal/cli/run.go (hunk 16 above)
Comment (ifireball) [reply internal/cli/run.go]: Good idea for a follow-up. Today security hooks live under workspace `.claude/hooks` (Claude Code contract) while plugin state lives under `CLAUDE_CONFIG_DIR` — see comment above `installClaudeHooks`. Unifying behind a generic hooks dir would need a design pass with Claude Code’s settings layout.

File: internal/runtime/claude.go (hunk 17 above)
Comment (ifireball) [reply internal/runtime/claude.go]: Fixed in 8bf10e4c: added `WorkspaceDir()` to `Runtime`; `ClearIterationArtifacts` uses `r.WorkspaceDir()` and `r.ConfigDir()`.

File: internal/runtime/claude.go (hunk 18 above)
Comment (ifireball) [reply internal/runtime/claude.go]: Fixed in 8bf10e4c: `ExtractTranscripts` uses `r.ConfigDir()` instead of the sandbox constant.

**Comment by ifireball** [comment]
All review threads from the first pass are resolved, and follow-ups from @maruiz93 / @rh-hemartin / fullsend-ai-review are addressed in `8bf10e4c` (bootstrap path validation, `WorkspaceDir()`, unexported `sanitizeOutput`, merge with `main`).

**Still blocking merge:** `CHANGES_REQUESTED` from @ralphbean — could you take another look and re-approve when satisfied? @waynesun09 same if anything remains on your side.

TOCTOU hardening remains tracked in #1801 (accepted risk for this PR).

File: internal/cli/run.go (hunk 16 above)
Comment (ifireball) [reply internal/cli/run.go]: Good catch — you are right.

`bootstrapCommon` is runner-level setup (fullsend binary, `.env.d`, `.security` for sandbox `scan context`). The `%s/.claude/hooks` mkdir was leftover from before the runtime split: security hook scripts are only installed when `ClaudeRuntime.Bootstrap` type-asserts `ClaudeHooksBootstrap` and calls `installClaudeHooks`.

Those hooks intentionally live under **workspace** `.claude/hooks` (and `workspace/.claude/settings.json`), not under `CLAUDE_CONFIG_DIR`, because that is where Claude Code reads Pre/PostToolUse hook wiring — separate from plugin marketplace files under config dir.

Latest push removes `.claude/hooks` from `bootstrapCommon` and creates the dir inside `installClaudeHooks` only when hooks are actually installed. A generic runtime-agnostic hooks folder is still a reasonable follow-up once we have a second backend.

**Review by ralphbean — APPROVED** [review]

File: docs/runtimes.md (hunk 15 above)
Comment (rh-hemartin) [reply docs/runtimes.md:34]: The idea would be that this is not even part of the runtime, but enforced from the parent layer, so security is not depending on the runtime. It is fine the way it is, I will bring this up in next work on this area.

**Review by rh-hemartin — APPROVED** [review]
